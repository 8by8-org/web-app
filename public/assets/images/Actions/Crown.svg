<svg width="24" height="18" viewBox="0 0 24 18" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M19.418 15.7627L21.1494 8.74023C22.4414 8.74023 23.5312 7.65918 23.5312 6.31445C23.5312 4.9873 22.4414 3.89746 21.1055 3.89746C19.7695 3.89746 18.6885 4.9873 18.6885 6.31445C18.6885 6.58691 18.7324 6.8418 18.8115 7.0791L16.7197 8.37109C16.4473 8.53809 16.2803 8.49414 16.1309 8.2832L13.3975 4.60059C14.0127 4.16113 14.417 3.44043 14.417 2.62305C14.417 1.2959 13.3271 0.206055 12 0.206055C10.6553 0.206055 9.57422 1.28711 9.57422 2.62305C9.57422 3.44043 9.97852 4.16113 10.6025 4.60059L7.86035 8.29199C7.71973 8.48535 7.56152 8.5293 7.30664 8.37109L5.18848 7.06152C5.25879 6.82422 5.30273 6.57812 5.30273 6.31445C5.30273 4.9873 4.22168 3.89746 2.88574 3.89746C1.54102 3.89746 0.459961 4.9873 0.459961 6.31445C0.459961 7.6416 1.53223 8.72266 2.85059 8.74023L4.57324 15.7627C4.93359 17.2217 5.72461 17.9512 7.23633 17.9512H16.7637C18.2754 17.9512 19.0664 17.2217 19.418 15.7627ZM12 3.50195C11.5166 3.50195 11.1211 3.10645 11.1211 2.62305C11.1211 2.13086 11.5166 1.75293 12 1.75293C12.4746 1.75293 12.8701 2.14844 12.8701 2.62305C12.8701 3.10645 12.4834 3.50195 12 3.50195ZM6.71777 9.88281C7.4209 10.2959 8.41406 10.252 9.0293 9.42578L12 5.42676L14.9971 9.45215C15.5947 10.2695 16.5791 10.3135 17.2383 9.91797L19.5234 8.53809L18.4248 13.0293H5.56641L4.46777 8.51172L6.71777 9.88281ZM2.88574 7.19336C2.40234 7.19336 2.00684 6.79785 2.00684 6.31445C2.00684 5.83105 2.40234 5.44434 2.88574 5.44434C3.36035 5.44434 3.75586 5.83984 3.75586 6.31445C3.75586 6.79785 3.36914 7.19336 2.88574 7.19336ZM21.1055 7.19336C20.6133 7.19336 20.2354 6.79785 20.2354 6.31445C20.2354 5.83984 20.6309 5.44434 21.1055 5.44434C21.5889 5.44434 21.9844 5.83105 21.9844 6.31445C21.9844 6.79785 21.5889 7.19336 21.1055 7.19336ZM7.21875 16.334C6.64746 16.334 6.30469 16.0703 6.18164 15.5342L5.96191 14.6553H18.0293L17.8184 15.5342C17.6865 16.0703 17.3525 16.334 16.7725 16.334H7.21875Z" fill="white"/>
</svg>
