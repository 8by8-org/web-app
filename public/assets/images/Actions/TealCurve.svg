<svg width="139" height="178" viewBox="0 0 139 178" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M35.9943 23.5898C46.8547 -0.0170483 63.8234 1.98319 77.0774 0.917595C102.177 1.10112 115.82 13.0538 131.029 35.4769C146.237 57.9001 138.693 81.682 116.444 106.83C91.3591 135.184 119.238 125.964 102.781 157.568C82.4765 196.56 22.5082 171.634 8.40257 141.571C-12.347 97.3476 13.6651 88.0138 24.0344 77.7799C44.7236 57.3607 24.6034 48.3498 35.9943 23.5898Z" fill="#02DDC3"/>
</svg>
