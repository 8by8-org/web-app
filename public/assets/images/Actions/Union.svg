<svg width="375" height="31" viewBox="0 0 375 31" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M86.3715 11.3361C32.9861 10.9505 15.191 3.39312 0 0H375V29.4584C375 29.4584 295.796 35.0879 249.566 24.8314C208.26 15.6672 180.122 12.0131 86.3715 11.3361Z" fill="white"/>
</svg>
