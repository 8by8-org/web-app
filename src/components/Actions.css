#button-style {
  color: black;
  background: #bfbfbf;
  border-radius: 0;
  border-color: #bfbfbf;
}
