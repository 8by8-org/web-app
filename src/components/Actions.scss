@import "./../styles/Components.scss";

@mixin blob-background(
  $background: url(../assets/images/Actions/TealCurve.svg)
) {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  height: 150px;
  width: 300px;
  background-repeat: no-repeat;
  background-position: center;
  text-align: center;
  font-family: $bebas-neue;
  line-height: 77px;
  letter-spacing: 0.03em;
  background-image: $background;
}

.main-content {
  background-color: $white;
  color: $black;

  .primary-button {
    @include button;
    outline: 4px solid white;
  }

  .secondary-button {
    @include secondary-button;
  }

  .avatar-and-status {
    display: flex;
    padding: 50px;
    position: relative;
  }

  .action-status {
    width: 55%;
  }

  .avatar-container {
    @include blob-background;
    width: 100%;
    height: auto;
    #challenger-avatar {
      width: 100%;
    }
    #challenger-name {
      font-size: 30px;
    }
  }

  #action-no-challenger {
    @include heading-text($black, 36px);
    padding: 30px;
  }

  .heading {
    @include heading-text($black, 36px);
    margin: 0px 0px 0px 0px;
    letter-spacing: 0.03em;
    .underline {
      @include text-underline;
    }
  }

  .subheading {
    @include heading-text($black, 24px);
    margin: 30px 0px 15px 30px;
    line-height: 29px;
    letter-spacing: 0.03em;
  }

  .text {
    font-family: $lato;
    font-size: 18px;
    line-height: 22px;
    letter-spacing: 0.02em;
    margin: 0;
  }

  .image {
    display: block;
    margin: auto;
  }

  .action-items {
    background: $black;
    padding: 35px;
  }

  .curve {
    background-color: $black;
    width: 100%;
  }
}
