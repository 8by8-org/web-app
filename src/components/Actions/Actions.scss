@import "./../../styles/Components.scss";

@mixin blob-background(
  $background: url(./../../assets/images/Actions/TealCurve.svg)
) {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  height: 150px;
  width: 300px;
  background-repeat: no-repeat;
  background-position: center;
  text-align: center;
  font-family: $bebas-neue;
  line-height: 77px;
  letter-spacing: 0.03em;
  background-image: $background;
}

.actions {
  background-color: $white;
  color: $black;

  .top {
    .avatar-and-status {
      display: flex;
      position: relative;
      padding: 0px 50px;
    }

    .action-status {
      width: 50%;
      font-size: 112px;
      padding-top: 10%;
    }

    .avatar-and-status-finished {
      display: flex;
      padding: 0px 50px;
      position: relative;
      flex-direction: column;
    }

    .action-status-finished {
      width: 100%;
      padding-top: 10%;
    }

    .avatar-container {
      @include blob-background;
      width: 100%;
      height: auto;
      padding: 10% 0% 10%;
      .images {
        position: relative;
      }
      #challenger-avatar {
        width: 100%;
        position: relative;
        top: 15px;
      }
      #challenger-crown {
        background-color: black;
        width: 30px;
        height: 30px;
        display: flex;
        justify-content: center;
        align-items: center;
        border-radius: 50%;
        position: absolute;
        bottom: -10px;
        right: 20px;
      }
      #challenger-name {
        @include heading-text($black, 30px);
      }
    }

    #action-no-challenger {
      @include heading-text($black, 36px);
      padding: 30px;
    }
  }

  .curve {
    background-color: $black;
    width: 100%;
  }

  .heading {
    @include heading-text($black, 48px);
    letter-spacing: 0.03em;
  }

  .subheading {
    @include subheading-text($white, 24px);
    margin: 0px 0px 30px;
    line-height: 29px;
    letter-spacing: 0.03em;
  }

  .text {
    @include text($white, 18px);
    line-height: 22px;
    letter-spacing: 0.02em;
    padding: 30px 0px 10px 0px;
    margin: 0px;
  }

  .action-items {
    background: $black;
    padding: 35px;
    color: white;
    text-align: center;

    .primary-button {
      @include button;
      // outline: 4px solid white;
    }

    .secondary-button {
      @include inverted-button;
      width: 100%;
    }
  }

  .links-container {
    display: flex;
    text-align: center;
    flex-direction: column;
  }

  .links {
    font-family: $lato;
    font-size: 18px;
    line-height: 22px;
    letter-spacing: 0.02em;
    color: $blue;
    padding-bottom: 10px;
  }

  .link-share {
    background-color: transparent;
    border: none;
    cursor: pointer;
    text-decoration: underline;
    display: inline;
    margin: 0;
    padding: 0px 0px 10px 0px;
    font-family: $lato;
    font-size: 18px;
    line-height: 22px;
    letter-spacing: 0.02em;
    color: $blue;
    font-weight: 900;
  }
}
