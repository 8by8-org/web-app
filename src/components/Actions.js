import React, { useState, useEffect } from 'react'
import { Button } from "react-bootstrap";
import { useHistory } from 'react-router'
import Avatar1 from '../assets/avatars/avatar1.svg'
import Avatar2 from '../assets/avatars/avatar2.svg'
import Avatar3 from '../assets/avatars/avatar3.svg'
import Avatar4 from '../assets/avatars/avatar4.svg'
import WhiteCurve from '../assets/images/Actions/Union.svg'
import './Actions.scss'

const avatars = [Avatar1, Avatar2, Avatar3, Avatar4]

export default function Actions() {
    const history = useHistory();
    const [ loading, setLoading ] = useState(true);
    const [ challengerInfo, setChallengerInfo ] = useState(null);
    
    useEffect(() => {
        if(localStorage.getItem("challengerInfo")) {
            setChallengerInfo(JSON.parse(localStorage.getItem("challengerInfo")))
        } 
        setLoading(false)
      }, [loading]);

    return (
        loading === false?
        <>
        <div className="main-content">
            <div className="top">
                { challengerInfo ? (
                    <>
                        <div className="avatar-and-status">
                            <div className="action-status">
                                <h1>Take Action For:</h1>
                            </div>
                            <div className="avatar-container">
                                <img 
                                    src={challengerInfo.avatar ? avatars[(challengerInfo.avatar)-1] : avatars[0]} 
                                    id="challenger-avatar"
                                />
                                <p id="challenger-name">{challengerInfo.name}</p>
                            </div>
                        </div>
                    </>
                    ) : <h1 id="action-no-challenger" align="center">Take Action</h1>
                }
            </div>
            <img src={WhiteCurve} className="curve"/> 
            <div className="action-items">
                <div className="py-2">
                    <Button className="primary-button" onClick={() => {history.push(`/voterreg`); localStorage.setItem('player', 'voterreg')}}>Register to vote</Button>
                </div>
                <div className="py-2">
                    <Button className="secondary-button" onClick={() => {history.push(`/election-reminders`); localStorage.setItem('player', 'election-reminders')}}>Get election reminders</Button>
                </div>
                <div className="py-2">
                    <Button className="secondary-button" onClick={() => {history.push('/challengerwelcome'); localStorage.setItem('player', 'progress')}}>Take the challenge</Button>
                </div>
            </div>
        </div>
        </>: <h1>loading</h1>
    )
}