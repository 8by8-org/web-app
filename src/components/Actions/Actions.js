import React, { useState, useEffect } from "react";
import { getUserDatabase } from "./../../functions/UserData";
import { Button } from "react-bootstrap";
import { useHistory } from "react-router";
import { useAuth } from "../../contexts/AuthContext";
import { getFirestore, getDoc, doc } from "firebase/firestore";
import Avatar1 from "./../../assets/avatars/avatar1.svg";
import Avatar2 from "./../../assets/avatars/avatar2.svg";
import Avatar3 from "./../../assets/avatars/avatar3.svg";
import Avatar4 from "./../../assets/avatars/avatar4.svg";
import WhiteCurve from "./../../assets/images/Actions/Union.svg";
import Crown from "./../../assets/images/Actions/Crown.svg";
import ConfettiAnimation from "./../ConfettiAnimation";
import Invite from "../Invite.js";
import "./Actions.scss";
import { LoadingWheel } from "../LoadingWheel/LoadingWheel.component";

const avatars = [Avatar1, Avatar2, Avatar3, Avatar4];

export default function Actions() {
  const history = useHistory();
  const { currentUser } = useAuth();
  const [loading, setLoading] = useState(false);
  const [challengerInfo, setChallengerInfo] = useState(null);
  const [registeredVoter, setRegisteredVoter] = useState(false);
  const [notifyElectionReminders, setNotifyElectionReminders] = useState(false);
  const [startedChallenge, setStartedChallenge] = useState(false);
  const toggleInvite = React.useRef();

  useEffect(() => {
    if (localStorage.getItem("challengerInfo")) {
      setChallengerInfo(JSON.parse(localStorage.getItem("challengerInfo")));
      if (currentUser) {
        fetchUserData();
      } else {
        setLoading(true);
      }
      setLoading(false);
    } else {
      if (currentUser) {
        getUserDatabase()
          .then((data) => {
            if (data.invitedBy) {
              setRegisteredVoter(data.isRegisteredVoter);
              setStartedChallenge(data.startedChallenge);
              setNotifyElectionReminders(data.notifyElectionReminders);
              getChallengerInfo(data.invitedBy);
              setLoading(false);
            } else {
              history.push(`/signin`);
            }
          })
          .catch((e) => console.log(e));
      } else {
        history.push(`/signin`);
      }
    }
  }, []);

  function fetchUserData() {
    getUserDatabase()
      .then((data) => {
        setRegisteredVoter(data.isRegisteredVoter);
        setStartedChallenge(data.startedChallenge);
        setNotifyElectionReminders(data.notifyElectionReminders);
        setLoading(true);
      })
      .catch((e) => console.log(e));
  }

  async function getChallengerInfo(invitedBy) {
    const db = getFirestore();
    const docRef = doc(db, "users", invitedBy);
    const query = await getDoc(docRef);
    const info = (({ name, avatar }) => ({ name, avatar }))(query.data());
    info.challengerID = invitedBy;
    setChallengerInfo(JSON.stringify(info));
  }

  return loading === false ? (
    <div>
      {/* if all three actions are completed */}
      {registeredVoter && notifyElectionReminders && startedChallenge ? (
        <div className="actions">
          <ConfettiAnimation time={8000} />

          <div className="top">
            <div className="avatar-and-status-finished">
              <div className="action-status-finished">
                <h1 className="heading">
                  You are done!
                  <br />
                  You Supported:
                </h1>
              </div>

              <div className="avatar-container">
                <div className="images">
                  <img
                    alt="Challenger Avatar"
                    src={
                      challengerInfo.avatar
                        ? avatars[challengerInfo.avatar - 1]
                        : avatars[0]
                    }
                    id="challenger-avatar"
                  />
                </div>

                <p id="challenger-name">{challengerInfo.name}</p>
              </div>
            </div>
          </div>

          <img alt="White Curve" src={WhiteCurve} className="curve" />

          <div className="action-items">
            <div className="py-2">
              <Button
                className="primary-button"
                onClick={() => {
                  history.push(`/signin`);
                }}
              >
                See Your Challenge
              </Button>
            </div>

            <div className="py-2">
              <Button
                className="secondary-button"
                onClick={() => toggleInvite.current()}
              >
                Share About Your Actions
              </Button>
            </div>
          </div>
        </div>
      ) : (
        // when there are actions that can still be completed by the user
        <div className="actions">
          <div className="top">
            <div className="avatar-and-status">
              <div className="action-status">
                {registeredVoter ||
                notifyElectionReminders ||
                startedChallenge ? (
                  <h1 className="heading">You Took Action!</h1>
                ) : (
                  <h1 className="heading">Take Action For:</h1>
                )}
              </div>

              <div className="avatar-container">
                <div className="images">
                  <img
                    alt="Challenger Avatar"
                    src={
                      challengerInfo.avatar
                        ? avatars[challengerInfo.avatar - 1]
                        : avatars[0]
                    }
                    id="challenger-avatar"
                  />
                  {(registeredVoter ||
                    notifyElectionReminders ||
                    startedChallenge) && (
                    <div id="challenger-crown">
                      <img
                        alt="Challenger Crown"
                        src={Crown}
                        className="crown"
                      />
                    </div>
                  )}
                </div>

                <p id="challenger-name">{challengerInfo.name}</p>
              </div>
            </div>
          </div>

          <img alt="White Curve" src={WhiteCurve} className="curve" />

          <div className="action-items">
            {/* when only Take the challenge action is left */}
            {registeredVoter && notifyElectionReminders && !startedChallenge ? (
              <h6 className="subheading">
                OTHER ACTIONS YOU CAN TAKE TO
                <br />
                HELP THE AAPI COMMUNITY
              </h6>
            ) : (
              // when there is at least one action completed
              (registeredVoter ||
                notifyElectionReminders ||
                startedChallenge) && (
                <h6 className="subheading">
                  {challengerInfo.name} GOT A BADGE!
                  <br />
                  HERE ARE OTHER ACTIONS TO HELP
                  <br />
                  THE AAPI COMMUNITY
                  <br />
                </h6>
              )
            )}

            {/* action buttons only displayed when they are not completed */}
            {!registeredVoter && (
              <div className="py-2">
                <Button
                  className="primary-button"
                  onClick={() => {
                    history.push(`/voterreg`);
                    localStorage.setItem("player", "voterreg");
                  }}
                >
                  Register to vote
                </Button>
              </div>
            )}

            {!notifyElectionReminders && (
              <div className="py-2">
                <Button
                  className={
                    registeredVoter ? "primary-button" : "secondary-button"
                  }
                  onClick={() => {
                    history.push(`/election-reminders`);
                    localStorage.setItem("player", "election-reminders");
                  }}
                >
                  Get election reminders
                </Button>
              </div>
            )}

            {!startedChallenge && (
              <div className="py-2">
                <Button
                  className={
                    registeredVoter && notifyElectionReminders
                      ? "primary-button"
                      : "secondary-button"
                  }
                  onClick={() => {
                    history.push("/challengerwelcome");
                    localStorage.setItem("player", "progress");
                  }}
                >
                  Take the challenge
                </Button>
              </div>
            )}

            {registeredVoter && notifyElectionReminders && !startedChallenge ? (
              // when only Take the challenge action is left
              <div>
                <h6 className="text">
                  Voting matters towards better AAPI
                  <br />
                  representation in our country. You can help
                  <br />
                  the community by encouring others to
                  <br />
                  register! Take the challenge on your own!
                  <br />
                </h6>
                <div className="links-container">
                  <a href="/homepage" className="links">
                    What is 8by8 Challenge?
                  </a>
                  <button
                    type="button"
                    className="link-share"
                    onClick={() => toggleInvite.current()}
                  >
                    Share about your actions
                  </button>
                </div>
              </div>
            ) : (
              // when at least one action is completed
              <div>
                {/* these are for when only that specific action is completed */}
                {registeredVoter &&
                  !notifyElectionReminders &&
                  !startedChallenge && (
                    <h6 className="text">Thanks for registering to vote!</h6>
                  )}

                {!registeredVoter &&
                  notifyElectionReminders &&
                  !startedChallenge && (
                    <h6 className="text">
                      Thanks for getting election reminders!
                    </h6>
                  )}

                {!registeredVoter &&
                  !notifyElectionReminders &&
                  startedChallenge && (
                    <h6 className="text">Thanks for taking the challenge!</h6>
                  )}

                {/* this is for when registered to vote or election reminders are turned on and challenge has been started */}
                {(registeredVoter || notifyElectionReminders) &&
                  startedChallenge && (
                    <h6 className="text">Thanks for your actions!</h6>
                  )}

                <div className="links-container">
                  {/* this is for when registered to vote or election reminders are turned on */}
                  {((registeredVoter &&
                    !notifyElectionReminders &&
                    !startedChallenge) ||
                    (!registeredVoter &&
                      notifyElectionReminders &&
                      !startedChallenge)) && (
                    <button
                      type="button"
                      className="link-share"
                      onClick={() => toggleInvite.current()}
                    >
                      Share about your action
                    </button>
                  )}

                  {/* this is for when registered to vote or election reminders are turned on and challenge has been started */}
                  {(registeredVoter || notifyElectionReminders) &&
                    startedChallenge && (
                      <button
                        type="button"
                        className="link-share"
                        onClick={() => toggleInvite.current()}
                      >
                        Share about your actions
                      </button>
                    )}

                  {/* this is for when user has started their own challenge */}
                  {startedChallenge && (
                    <a href="/signin" className="links">
                      See your challenge
                    </a>
                  )}
                </div>
              </div>
            )}
          </div>
        </div>
      )}

      <Invite toggleInvite={toggleInvite} isShare={true} />
    </div>
  ) : (
    <LoadingWheel overlay={false} />
  );
}
