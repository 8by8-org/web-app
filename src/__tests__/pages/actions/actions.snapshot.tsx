import Actions from '@/pages/actions';
import { cleanup, render } from '@testing-library/react';

import { useRouter } from 'next/router';

jest.mock('next/router', () => ({
  useRouter: jest.fn()
}))

describe('Actions', () => {
  afterEach(cleanup);
  
  it('renders actions page unchanged', () => {
    const { container } = render(<Actions />); 
    expect(container).toMatchSnapshot();
  });
});