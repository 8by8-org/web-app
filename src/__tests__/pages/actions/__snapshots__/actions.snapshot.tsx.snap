// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Actions renders actions page unchanged 1`] = `<div />`;
